Added delayed command execution system

Brownshome opened the pull request:
Adds commands and a task to handle execution of delayed commands using the RTC. Several additional functions and error handling were also added during the construction of this module.

The system maintains a file `delay.txt` that list delayed commands. Commands are removed from this file as they are executed and queued onto the file.

The task checks the file every 1s and executes commands that are scheduled for this tick.

Possible changes:
 - Store the command file in a text-only format, currently the date and time-stamps are dumped to file in their BCD format
 - Sleep the task more intelligently based on the delay to the next call
 - Use seconds since and epoch instead of `Y:M:D:H:M:S` stamps
 - Nuances in error-handling, currently the task exits under most file-system errors. A better solution may be to revert to an in-memory backup if the SD card is lost.

Brownshome committed:
Added delayed command execution system

Brownshome committed:
Renamed command file

ConnorMattson commented on src/Shared.c:
```diff
@@ -0,0 +1,33 @@
+
+#include "MCU_Init.h"
+
+HAL_StatusTypeDef EXT(CORE_getTime)(RTC_TimeTypeDef *time, RTC_DateTypeDef *date, uint32_t format) {
```
This is very different from EnduroSat's implementation: [Link to Koios OBC since this isnt implemented in Kessler's version](https://github.com/APSS-Koios/apss3-obc/blob/4c764ae84bb48462e030a7fe35faf1f08d98dfd8/Middlewares/ESTTC.c#L1124)

Brownshome replied:
Yeah I did see this. @Evander7 also had a similar implementation. The synchro appears to only be needed if the time had recently been updated or initialised. So I am not sure if it would be needed. The delay also seems unneeded, I can see no reference anywhere to a delay requirement in any document for the HAL interface; and I would assume that any delay requirements have already been met if needed by a synchro call:

From the HAL source:
> To read the calendar through the shadow registers after Calendar 
initialization, calendar update or after wake-up from low power modes 
the software must first clear the RSF flag. The software must then 
wait until it is set again before reading the calendar, which means 
that the calendar registers have been correctly copied into the 
RTC_TR and RTC_DR shadow registers.The HAL_RTC_WaitForSynchro() function 
implements the above software sequence (RSF clear and RSF check).

Additionally, `HAL_RTC_WaitForSynchro` requires write protection be dissabled on the RTC registers. Which does not appear to happen in the endurosat code at any point, so I'm not sure the call will have any effect anyway.

I would be interested in testing if this is the case. But for now I think leaving the code as it is is the best bet, as this seems to be the correct implementation as per the spec.

ConnorMattson commented on src/DelayedCommand.c:
```diff
@@ -0,0 +1,253 @@
+    EXT(CORE_exitSDRegion)();
+}
+
+static void taskFunction(void const *arg) {
```
Without calls to the task monitor this task will die.

Brownshome replied:
Yeah, I was not sure about this. My current working (untested) theory, is that if the task monitor is not told about a particular task then it won't monitor it.

Brownshome replied:
I'll leave this out until it is determined that we need it

ConnorMattson commented:
I think this code would be much better as a library rather than trying to implement a task. Allow the user to make use of this functionality in the way that they want. Don't try and force a specific use case upon them.

The mixture of a mail queue and file handling makes it very difficult to understand how this is intended to be used.

Brownshome commented:
I agree, I'll see if I can restructure it into a library. That might play badly with the way I've structured the inter-task communication. I'll see if I can make it easier to follow

Brownshome committed:
Added a non-task based code-path to aid KOIOS' use-case.

Brownshome commented:
I've added comments to all internal functions and added functions that should allow the system to be used as a library. If bits are still difficult to understand add specific comments and I'll document the code in those places more.

ConnorMattson reviewed (approved):
Thanks for making those changes, looks good!
